### PR TITLE
group norm with binf & sasha's upper bound

### PR DESCRIPTION
**Pull request by rjbaraldi**

### CODE REVIEW

File: src/shiftedGroupNormL2Binf.jl [hunk 1]
```diff
@@ -0,0 +1,99 @@
+export ShiftedGroupNormL2Binf
+
+mutable struct ShiftedGroupNormL2Binf{
+  R  <: Real,
+	RR <: AbstractVector{R},
```
Comment (dpo) [inline src/shiftedGroupNormL2Binf.jl:5]:
```suggestion
  RR <: AbstractVector{R},
```

File: src/shiftedGroupNormL2Binf.jl [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  shifted_twice::Bool
+
+  function ShiftedGroupNormL2Binf(
+    h::GroupNormL2{R,RR,I},
+    xk::AbstractVector{R},
+    sj::AbstractVector{R},
+    Δ::R,
+    χ::Conjugate{IndBallL1{R}},
+    shifted_twice::Bool,
+		) where {R <: Real, RR <: AbstractVector{R}, I}
```
Comment (dpo) [inline src/shiftedGroupNormL2Binf.jl:26]:
```suggestion
  ) where {R <: Real, RR <: AbstractVector{R}, I}
```

File: src/shiftedGroupNormL2Binf.jl [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  "xk = $(ψ.xk)\n" * " "^14 * "sj = $(ψ.sj)\n" * " "^14 * "Δ = $(ψ.Δ)"
+
+function prox!(
+  y::AbstractVector{R},
+  ψ::ShiftedGroupNormL2Binf{R, RR,  I, V0, V1, V2},
+  q::AbstractVector{R},
+  σ::R,
+ ) where {R <: Real, RR <: AbstractVector{R}, I, V0 <: AbstractVector{R}, V1 <: AbstractVector{R}, V2 <: AbstractVector{R}}
+
+  ψ.sol .= q + ψ.xk + ψ.sj
```
Comment (dpo) [inline src/shiftedGroupNormL2Binf.jl:64]:
```suggestion
  ψ.sol .= q .+ ψ.xk .+ ψ.sj
```

File: src/shiftedGroupNormL2Binf.jl [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  ϵ = 1 ## sasha's initial guess
+	softthres(x, a) = sign.(x) .* max.(0, abs.(x) .- a)
```
Comment (dpo) [inline src/shiftedGroupNormL2Binf.jl:66]:
```suggestion
  softthres(x, a) = sign.(x) .* max.(0, abs.(x) .- a)
```
This function creates a new array.

File: src/shiftedGroupNormL2Binf.jl [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+  l2prox(x,a) = max(0, 1 - a/norm(x)).*x
```
Comment (dpo) [inline src/shiftedGroupNormL2Binf.jl:67]: This one too.

File: src/shiftedGroupNormL2Binf.jl [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+  for (idx, λ) ∈ zip(ψ.h.idx, ψ.h.lambda)
+    σλ  = λ*σ
+    ## find root for each block
+    froot(n) = n - norm(σ .* softthres((ψ.sol[idx]./σ .- (n/(σ*(n - σλ))) .* ψ.xk[idx]), ψ.Δ*(n/(σ*(n - σλ)))) .- ψ.sol[idx] )
```
Comment (dpo) [inline src/shiftedGroupNormL2Binf.jl:71]:
```suggestion
    froot(n) = n - norm(σ .* softthres((ψ.sol[idx]./σ .- (n/(σ*(n - σλ))) .* ψ.xk[idx]), ψ.Δ*(n/(σ*(n - σλ)))) .- ψ.sol[idx])
```

File: src/shiftedGroupNormL2Binf.jl [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
```
Comment (dpo) [inline src/shiftedGroupNormL2Binf.jl:72]:
```suggestion
```

File: src/shiftedGroupNormL2Binf.jl [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+		lmin = σλ*(1+eps()) #lower bound
```
Comment (dpo) [inline src/shiftedGroupNormL2Binf.jl:73]:
```suggestion
  lmin = σλ*(1+eps(R)) # lower bound
```

File: src/shiftedGroupNormL2Binf.jl [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    fl = froot(lmin)
+
+    ansatz = lmin + ϵ #ansatz for upper bound
+    step = ansatz / (σ*(ansatz - σλ))
+    zlmax = norm(softthres((ψ.sol[idx]./σ .- step .* ψ.xk[idx]), ψ.Δ*step))
+    lmax = norm(ψ.sol[idx]) + σ*(zlmax + abs((ϵ - 1)/ϵ+1)*λ*norm(ψ.xk[idx]))
+ 	 	fm = froot(lmax)
```
Comment (dpo) [inline src/shiftedGroupNormL2Binf.jl:80]: Indentation all screwed up. You have mixtures of tabs and spaces here.

**Review by dpo — CHANGES_REQUESTED** [review]

File: src/shiftedGroupNormL2Binf.jl (hunk 5 above)
Comment (rjbaraldi) [reply src/shiftedGroupNormL2Binf.jl:67]:
```suggestion
  l2prox(x,a) = max(0, 1 - a/norm(x)).*x
```
Just two spaces?

File: src/shiftedGroupNormL2Binf.jl (hunk 8 above)
Comment (rjbaraldi) [reply src/shiftedGroupNormL2Binf.jl:72]: Idk what's up with all these tabs suddenly. Thanks for checking through them.

**Comment by dpo** [comment]
You've got tabs everywhere. What happened? Did you change the VSCode config?

**Pull request merged**